lib/eslint.js

```diff
@@ -807,20 +811,18 @@ module.exports = (function() {
              */
             controller.traverse(ast, {
                 enter: function(node, parent) {
-
                     var comments = api.getComments(node);
 
                     emitCommentsEnter(comments.leading);
                     node.parent = parent;
-                    api.emit(node.type, node);
+                    codePathAnalyzer.enterNode(node);
```

Comment:
I have three ideas.

1. To separate methods.

  ```js
  codePathAnalyzer.enterNode(node);
  api.emit(node.type, node);
  codePathAnalyzer.leaveEnteringNode(node);

  //....

  codePathAnalyzer.enterLeavingNode(node);
  api.emit(node.type + ":exit", node);
  codePathAnalyzer.leaveNode(node);
  ```

2. A callback.

  ```js
  codePathAnalyzer.enterNode(node, function() {
      api.emit(node.type, node);
  });

  //....

  codePathAnalyzer.leaveNode(node, function() {
      api.emit(node.type + ":exit", node);
  });
  ```

3. The decorator pattern. 

![image](https://cloud.githubusercontent.com/assets/1937871/9675875/652c78e0-52fe-11e5-9cf6-0c72a2ea8bc0.png)

  ```js
  var visitor = new CommentTraverser(new CodePathAnalyzer(new NodeTraverser(api), api), api);

  //....

  visitor.enterNode(node);

  //....

  visitor.leaveNode(node);
  ```